### PR TITLE
Make closures carry their own ClosureKind

### PR DESCRIPTION
**Pull request by compiler-errors**
Right now, we use the "`movability`" field of `hir::Closure` to distinguish a closure and a coroutine. This is paired together with the `CoroutineKind`, which is located not in the `hir::Closure`, but the `hir::Body`. This is strange and redundant.

This PR introduces `ClosureKind` with two variants -- `Closure` and `Coroutine`, which is put into `hir::Closure`. The `CoroutineKind` is thus removed from `hir::Body`, and `Option<Movability>` no longer needs to be a stand-in for "is this a closure or a coroutine".

r? eholk

### CODE REVIEW

File: compiler/rustc_hir/src/hir.rs [hunk 1]
```diff
@@ -3661,7 +3670,7 @@ mod size_asserts {
     use super::*;
     // tidy-alphabetical-start
     static_assert_size!(Block<'_>, 48);
-    static_assert_size!(Body<'_>, 32);
+    static_assert_size!(Body<'_>, 24);
```
Comment (compiler-errors) [inline compiler/rustc_hir/src/hir.rs:3673]: Yes, it makes `Body` smaller. I doubt that will have much of an effect, tho.

**Comment by compiler-errors** [comment]
I doubt this changes much, but 

@bors try @rust-timer queue

File: compiler/rustc_hir/src/hir.rs (hunk 1 above)
Comment (nnethercote) [reply compiler/rustc_hir/src/hir.rs:3673]: Probably not, but it's better than making it bigger :)

File: compiler/rustc_borrowck/src/diagnostics/conflict_errors.rs [hunk 2]
```diff
@@ -1338,7 +1346,7 @@ impl<'cx, 'tcx> MirBorrowckCtxt<'cx, 'tcx> {
                 //     mut iter => {
                 //         [opt_ident]: loop {
                 //             match Iterator::next(&mut iter) {
-                //                 None => break,
+                //                 hir::ClosureKind::Closure =>  break,
```
Comment (cjgillot) [inline compiler/rustc_borrowck/src/diagnostics/conflict_errors.rs:1349]: Spurious change.

File: compiler/rustc_borrowck/src/diagnostics/region_name.rs [hunk 3]
```diff
@@ -683,47 +683,49 @@ impl<'tcx> MirBorrowckCtxt<'_, 'tcx> {
...
-                                span = output.span();
-                                if let hir::FnRetTy::Return(ret) = output {
-                                    hir_ty = Some(self.get_future_inner_return_ty(*ret));
-                                }
-                                " of async function"
+                let mir_description = match kind {
+                    hir::ClosureKind::Coroutine(
+                        hir::CoroutineKind::Desugared(hir::CoroutineDesugaring::Async, src),
+                        _,
+                    ) => match src {
```
Comment (cjgillot) [inline compiler/rustc_borrowck/src/diagnostics/region_name.rs:690]: Can this be flattened to match on the desugaring and the source at the same time?

File: compiler/rustc_hir/src/hir.rs [hunk 4]
```diff
@@ -945,7 +945,21 @@ pub struct Closure<'hir> {
...
+pub enum ClosureKind {
+    /// This is a plain closure expression.
+    Closure,
+    /// This is a coroutine expression -- i.e. a closure expression in which
+    /// we've found a `yield`. These can arise either from "plain" coroutine
+    ///  usage (e.g. `let x = || { yield (); }`) or from a desugared expression
+    /// (e.g. `async` and `gen` blocks).
+    // FIXME(coroutines): We could probably remove movability here -- it can be deduced
+    // from the `CoroutineKind` in all cases (except for "plain" coroutines, which could
+    // carry the movability in the variant).
```
Comment (cjgillot) [inline compiler/rustc_hir/src/hir.rs:961]: Do you want to leave this as a follow-up, or do it here ?

File: compiler/rustc_borrowck/src/diagnostics/conflict_errors.rs (hunk 2 above)
Comment (compiler-errors) [reply compiler/rustc_borrowck/src/diagnostics/conflict_errors.rs:1349]: Find and replace error 🤦

File: compiler/rustc_hir/src/hir.rs (hunk 4 above)
Comment (compiler-errors) [reply compiler/rustc_hir/src/hir.rs:961]: Yeah, I can do it here. Shouldn't be too much more work.

File: compiler/rustc_ast_lowering/src/expr.rs [hunk 5]
```diff
@@ -1656,7 +1655,19 @@ impl<'hir> LoweringContext<'_, 'hir> {
...
                 );
             }
-            Some(hir::CoroutineKind::Coroutine) | None => {
+            Some(hir::CoroutineKind::Coroutine(_)) => {
+                if !self.tcx.features().coroutines {
+                    rustc_session::parse::feature_err(
+                        &self.tcx.sess.parse_sess,
+                        sym::coroutines,
+                        span,
+                        "yield syntax is experimental",
```
Comment (eholk) [inline compiler/rustc_ast_lowering/src/expr.rs:1664]: Huh, weird that this wasn't here earlier...

Comment (compiler-errors) [reply compiler/rustc_ast_lowering/src/expr.rs:1664]: It was there! I just split out the branches so the code got duplicated.

Comment (eholk) [reply compiler/rustc_ast_lowering/src/expr.rs:1664]: Oh, there it is, down below. Duh!

File: compiler/rustc_hir/src/hir.rs (hunk 4 above)
Comment (eholk) [inline compiler/rustc_hir/src/hir.rs:961]: This `FIXME` has been taken care of already, right?

Comment (compiler-errors) [reply compiler/rustc_hir/src/hir.rs:961]: Yes, oop

File: compiler/rustc_middle/src/middle/region.rs
Comment (eholk) [inline compiler/rustc_middle/src/middle/region.rs:1]: The changes in this file seem unrelated...

Comment (eholk) [reply compiler/rustc_middle/src/middle/region.rs:1]: I'm all for deleting code if it's not needed though!

Comment (compiler-errors) [reply compiler/rustc_middle/src/middle/region.rs:1]: I deleted `body_expr_count` because computing it requires knowing the coroutine kind of the body, which is now difficult to compute:

https://github.com/rust-lang/rust/pull/119258/files/ba912855cc1aff5f2b403411c8d890d3978abf9a#diff-c4f75d1b2e3c322eb44d7ee2d8e67b495c41441dfb3b96218c8759c2c2342f3eL827

**Comment by compiler-errors** [comment]
this is enough changes that i'd rather not roll this up

@bors rollup=never
